eip-0023/eip-0023.md

```diff
@@ -0,0 +1,337 @@
+The ballot token holders and oracle token holders need not be the same people. 
+The logic for updating the pool is stored in an **update box**, which identified by the **update-NFT**. This NFT is referenced in the pool and ballot boxes.
+The ballot box allows the owner to vote for 3 things: 
+
+1. The script of the new pool box.
+2. The reward token id to be stored in the new pool box.
+3. The reward token amounts to be stored in the new pool box.
+
+Apart from this, the new pool box *must* preserve the remaining values from the old pool box.    
+
+**Note** The update box logic acts on the pool box and not the refresh box.
+
+## Tokens
+
+The system has the following types of tokens. Note that we use the term **NFT** to refer to any token that was issued in quantity 1.
+
+| Token | Issued quantity | purpose | where stored |
+|---|---|---|---|
+|Pool-NFT | 1 | Identify pool box | Pool box |
+|Refresh-NFT | 1 | Identify refresh box | Refresh box |
+|Update-NFT | 1 | Identify update box | Update box |
+|Oracle tokens | 15 | Identify each oracle box | Oracle boxes |
+|Ballot tokens | 15 | Identify each ballot box | Ballot boxes |
+|Reward tokens | 100 million | Reward oracles | Pool box <br /> Oracle boxes |
+
+## Boxes
+
+There are a total of 5 contracts, each corresponding to a box type
+
+| Box | Quantity | Tokens | Additional registers used | Purpose | Spending transactions |
+|---|---|---|---|---|---|
+|Pool | 1 | Pool-NFT <br /> Reward tokens| R4: Rate (Long) <br /> R5: Epoch counter (Int) | Publish pool rate for dApps <br /> Emit reward tokens| Refresh pool <br /> Update pool | 
+|Refresh| 1 | Refresh-NFT | | Refresh pool box | Refresh pool |
+|Oracle | 15 | Oracle token <br /> Reward tokens | R4: Public key (GroupElement) <br /> R5: Epoch counter of pool box (Int) <br /> R6: Published rate (Long) | Publish data-point <br /> Accumulate reward tokens | Publish data-point, <br /> Refresh pool, <br /> Transfer oracle token, <br /> Extract reward tokens|
+|Update | 1 | Update-NFT | | Updating pool box | Update pool box |
+|Ballot | 15 | Ballot token | R4: Public key (GroupElement) <br /> R5: Update box creation height (Int) <br /> R6: New pool box hash (Coll[Byte]) | Voting for updating pool box | Vote for update <br /> Update pool box <br /> Transfer ballot token |
+
+Before going into the transactions in the protocol, we first present the contracts for each of the above boxes.
+
+## Contracts with base-64-encoded hash of ergo-tree bytes
+
+- [Pool Contract](contracts/pool_contract.es) `8cJi+FGGU32jXyO8M2LeyWSWlerdcb1zxBWeZtyy7Y8=`
+- [Refresh Contract](contracts/refresh_contract.es) `cs5c5QEirstI4ZlTyrbTjlPwWYHRW+QsedtpyOSBnH4=`
+- [Oracle Contract](contracts/oracle_contract.es) `fhOYLO3s+NJCqTQDWUz0E+ffy2T1VG7ZnhSFs0RP948=`
+- [Ballot Contract](contracts/ballot_contract.es) `2DnK+72bh+TxviNk8XfuYzLKtuF5jnqUJOzimt30NvI=`
+- [Update Contract](contracts/update_contract.es) `0wFmk/1TNpgTsbzpWND3WLPbwQdD8E+TWDzZLaYv3nE=`
```

Comment:
```suggestion
- [Update Contract](contracts/update_contract.es) `3aIVTP5tRgCZHxaT3ZFw3XubRV5DJi0rKeo9bKVHlVw=`
```
Generated from https://scastie.scala-lang.org/BxJFCRDcSCeiwf9ZgKdymQ
using the latest update contract https://github.com/ergoplatform/eips/blob/1757e0060d6dafa6bd9bcb2339ab5484dfc767a7/eip-0023/contracts/update_contract.es with pool box preservation requirement removed.
Corresponding PR in oracle-core is https://github.com/ergoplatform/oracle-core/pull/178